Refactor single user groups

dannygoldstein opened the pull request:
This PR refactors the single user group implementation. 

In this PR, single user groups are created, updated, and deleted automatically through event listeners. This removes the need to ever explicitly create a single user group (e.g., in a handler, or in a test fixture). 

The advantage of this approach is that one can rest assured that single user groups will always exist, for all users. In the current test suite, single user groups are not created automatically for all users, e.g., in fixtures. This creates situations where some users do not have associated single user groups, which breaks the ownership model for photometry and spectra (posted photometry and spectra should always be shared with the posting user's single user group by default).

dannygoldstein commented:
Failure unrelated

acrellin commented:
The reason I didn't use event listeners to begin with was because of SQLa's warnings that they should only be used in a very limited way, e.g. augmenting values on the row in question

dannygoldstein commented:
That is only for mapper level events, here the action is all happening in flush events, which are much safer - 

> Probably the most widely used series of events are the “persistence” events, which correspond to the flush process. The flush is where all the decisions are made about pending changes to objects and are then emitted out to the database in the form of INSERT, UPDATE, and DELETE statements.
> 
> before_flush()
> The SessionEvents.before_flush() hook is by far the most generally useful event to use when an application wants to ensure that additional persistence changes to the database are made when a flush proceeds. Use SessionEvents.before_flush() in order to operate upon objects to validate their state as well as to compose additional objects and references before they are persisted. __Within this event, it is safe to manipulate the Session’s state, that is, new objects can be attached to it, objects can be deleted, and individual attributes on objects can be changed freely, and these changes will be pulled into the flush process when the event hook completes.__
> 
> The typical SessionEvents.before_flush() hook will be tasked with scanning the collections Session.new, Session.dirty and Session.deleted in order to look for objects where something will be happening.
> 
> For illustrations of SessionEvents.before_flush(), see examples such as Versioning with a History Table and Versioning using Temporal Rows.
> 
> after_flush()
> The SessionEvents.after_flush() hook is called after the SQL has been emitted for a flush process, but before the state of the objects that were flushed has been altered. That is, you can still inspect the Session.new, Session.dirty and Session.deleted collections to see what was just flushed, and you can also use history tracking features like the ones provided by AttributeState to see what changes were just persisted. In the SessionEvents.after_flush() event, additional SQL can be emitted to the database based on what’s observed to have changed.
> 
> after_flush_postexec()
> SessionEvents.after_flush_postexec() is called soon after SessionEvents.after_flush(), but is invoked after the state of the objects has been modified to account for the flush that just took place. The Session.new, Session.dirty and Session.deleted collections are normally completely empty here. Use SessionEvents.after_flush_postexec() to inspect the identity map for finalized objects and possibly emit additional SQL. __In this hook, there is the ability to make new changes on objects, which means the Session will again go into a “dirty” state; the mechanics of the Session here will cause it to flush again if new changes are detected in this hook if the flush were invoked in the context of Session.commit(); otherwise, the pending changes will be bundled as part of the next normal flush.__ When the hook detects new changes within a Session.commit(), a counter ensures that an endless loop in this regard is stopped after 100 iterations, in the case that an SessionEvents.after_flush_postexec() hook continually adds new state to be flushed each time it is called.
> 
>

acrellin commented:
> That is only for mapper level events, here the action is all happening in flush events, which are much safer -
> 
> > Probably the most widely used series of events are the “persistence” events, which correspond to the flush process. The flush is where all the decisions are made about pending changes to objects and are then emitted out to the database in the form of INSERT, UPDATE, and DELETE statements.
> > before_flush()
> > The SessionEvents.before_flush() hook is by far the most generally useful event to use when an application wants to ensure that additional persistence changes to the database are made when a flush proceeds. Use SessionEvents.before_flush() in order to operate upon objects to validate their state as well as to compose additional objects and references before they are persisted. **Within this event, it is safe to manipulate the Session’s state, that is, new objects can be attached to it, objects can be deleted, and individual attributes on objects can be changed freely, and these changes will be pulled into the flush process when the event hook completes.**
> > The typical SessionEvents.before_flush() hook will be tasked with scanning the collections Session.new, Session.dirty and Session.deleted in order to look for objects where something will be happening.
> > For illustrations of SessionEvents.before_flush(), see examples such as Versioning with a History Table and Versioning using Temporal Rows.
> > after_flush()
> > The SessionEvents.after_flush() hook is called after the SQL has been emitted for a flush process, but before the state of the objects that were flushed has been altered. That is, you can still inspect the Session.new, Session.dirty and Session.deleted collections to see what was just flushed, and you can also use history tracking features like the ones provided by AttributeState to see what changes were just persisted. In the SessionEvents.after_flush() event, additional SQL can be emitted to the database based on what’s observed to have changed.
> > after_flush_postexec()
> > SessionEvents.after_flush_postexec() is called soon after SessionEvents.after_flush(), but is invoked after the state of the objects has been modified to account for the flush that just took place. The Session.new, Session.dirty and Session.deleted collections are normally completely empty here. Use SessionEvents.after_flush_postexec() to inspect the identity map for finalized objects and possibly emit additional SQL. **In this hook, there is the ability to make new changes on objects, which means the Session will again go into a “dirty” state; the mechanics of the Session here will cause it to flush again if new changes are detected in this hook if the flush were invoked in the context of Session.commit(); otherwise, the pending changes will be bundled as part of the next normal flush.** When the hook detects new changes within a Session.commit(), a counter ensures that an endless loop in this regard is stopped after 100 iterations, in the case that an SessionEvents.after_flush_postexec() hook continually adds new state to be flushed each time it is called.

@dannygoldstein okay, awesome

dannygoldstein commented on skyportal/models.py:
```diff
@@ -267,6 +270,15 @@ class Stream(Base):
+    lambda self: DBSession()
+    .query(Group)
+    .join(GroupUser)
+    .filter(Group.single_user_group.is_(True), GroupUser.user_id == self.id)
```
This should uniquely specify the relevant single user Group, I think.

dannygoldstein commented on skyportal/models.py:
```diff
@@ -2633,4 +2645,32 @@ def send_source_notification(mapper, connection, target):
 
+@event.listens_for(User, 'after_insert')
+def create_single_user_group(mapper, connection, target):
+    # Create single-user group
+
+    @event.listens_for(DBSession(), "after_flush", once=True)
```
fixed

dannygoldstein commented on skyportal/models.py:
```diff
@@ -2633,4 +2645,32 @@ def send_source_notification(mapper, connection, target):
+def update_single_user_group(mapper, connection, target):
+    @event.listens_for(DBSession(), "after_flush_postexec", once=True)
+    def receive_after_flush(session, context):
+
```
fixed

dannygoldstein commented:
failures unrelated - @stefanv ready for you to look at

stefanv reviewed (commented):
I know @acrellin  has approved, but I'm going to have him make the final decision on merging too.

It is not entirely clear to me what the risks are associated with listeners; the implementation seems cleaner, but I am simply not qualified to judge.

dannygoldstein commented:
failures unrelated
